gcs/CONFIGURATION.md

```diff
@@ -131,6 +131,10 @@
     fs.gs.storage.http.headers.another-custom-header=another_custom_value
     ```
 
+*   `fs.gs.audit.logs.header.enable` (default: `false`)
```

Comment:
Is there are reason we need to add this functionality in the GCE connector itself? We intentionally added support for generic HTTP headers so we do not need to implement per-header logic in GCS connector.

Reply:
This is added for the case when user submits the job through open source API and does not set audit log headers but sets this property as enabled. In that case we pick the Kerberos/OS level user and set that in the header. We are not doing any changes to the HTTP headers that the user has set using fs.gs.storage.http.headers

Reply:
Do we need to cover this though? For example, there will be no audit headers if users submit job that does not use GCS connector, or users gsutil. I think it's maybe OK to have a limitation that job should be submitted via Dataproc API to have this feature, especially taking into account that this is a temporary solution, IIUC.

Reply:
Thank you for the suggestion Igor. Meha, can we re-use the existing header for this?

Reply:
Btw, did we test if Hadoop's support for resolving env vars will work for us w/o any code changes?

I.e. if we will add this config to the Hadoop XML config (not sure about exact syntax, we may need to escape `$`):
```xml
<configuration>
  . . .
  <property>
    <name>fs.gs.storage.http.headers.X-Goog-Custom-Audit-User</name>
    <value>${user.name}</value>
  </property>
  . . .
</configuration>
```

If this works, then I think that we may not need even Dataproc Agent changes that sets this property manually for each user.

cc @cnauroth

Reply:
As per Javadoc in the [`Configuration`](https://github.com/apache/hadoop/blob/trunk/hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/conf/Configuration.java) class, configuration property expansion can pull from environment variables (prefixed as `${env.FOO}`), or other properties defined in the same file, or general JVM system properties. In the example above, using `${user.name}`, it would use the user name that launched the JVM process (e.g. Spark driver or YARN containers running Spark executors). In the context of Dataproc, the user running that JVM process varies depending on the cluster configuration and method of job submission:

| Cluster Configuration | Job Submission | User |
| - | - | - |
| Unsecured | API | yarn |
| Unsecured | ssh | yarn |
| Kerberos | API | dataproc |
| Kerberos | ssh | <user authenticated by `kinit`> |
| Personal Auth | API | <user authenticated by `kinit`> |
| Personal Auth | ssh | N/A, ssh blocked for Personal Auth clusters |

Does this satisfy the requirements? If so, then I expect you can make use of Hadoop configuration property expansion.

However, the earlier conversation discusses wanting to pick the "OS level user." If that is a requirement, then this solution would be insufficient, because:

1. In unsecured clusters, audit log headers would show the user as "yarn" instead of the submitting user.
2. In Kerberos clusters (but not using Personal Auth), for jobs submitted through the Dataproc API, audit log headers would show the user as "dataproc" instead of the submitting user.

If we need to satisfy that, then I can't think of any other solution other than the agent injecting a property.